File: cses/09_geometry/3427_line-segments-trace-i/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    # Quy mỗi đoạn về đường thẳng f(x) = s*x + b, với s là slope nguyên và
    # b = giá trị tại x = 0. Cùng một slope chỉ giữ intercept lớn nhất.
    best = {}
    for _ in range(n):
        y1 = int(data[idx]); y2 = int(data[idx + 1]); idx += 2
        s = (y2 - y1) // m          # slope nguyên, phép chia là chính xác
        b = y1                      # giá trị tại x = 0
        cur = best.get(s)
        if cur is None or b > cur:
            best[s] = b

    # Sắp xếp các đường theo slope tăng dần.
    lines = sorted(best.items())    # [(slope, intercept), ...]

    # Xây upper hull cho phép lấy MAX (Convex Hull Trick đơn điệu).
    hm = []   # danh sách slope
    hc = []   # danh sách intercept
    for s, c in lines:
        while len(hm) >= 2:
            m1 = hm[-2]; c1 = hc[-2]
            m2 = hm[-1]; c2 = hc[-1]
            # Loại l2 nếu giao(l2, l3) <= giao(l1, l2) (cross-multiplication nguyên):
            # (c2 - c)*(m2 - m1) <= (c1 - c2)*(s - m2)
            if (c2 - c) * (m2 - m1) <= (c1 - c2) * (s - m2):
                hm.pop(); hc.pop()
            else:
                break
        hm.append(s); hc.append(c)

    # Trả lời truy vấn x = 0..m bằng con trỏ tiến dần (chỉ số tối ưu không giảm).
    L = len(hm)
    out = []
    ptr = 0
    for x in range(m + 1):
        while ptr + 1 < L and hm[ptr + 1] * x + hc[ptr + 1] >= hm[ptr] * x + hc[ptr]:
            ptr += 1
        out.append(hm[ptr] * x + hc[ptr])

    sys.stdout.write(' '.join(map(str, out)))
    sys.stdout.write('\n')


main()

File: cses/09_geometry/3427_line-segments-trace-i/solution.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <vector>

using i128 = __int128_t;

// Mỗi đoạn được coi như một đường thẳng đầy đủ f(x) = slope*x + intercept.
struct Line {
    long long slope;
    long long intercept;

    long long value(const long long x) const {
        return slope * x + intercept;
    }
};

// Kiểm tra đường "middle" có vô dụng trên bao trên hay không.
// Dùng cross-multiplication trên __int128 để tránh chia và tránh tràn số:
// giao điểm (middle, last) không nằm phải hơn giao điểm (first, middle).
bool redundant(const Line& first, const Line& middle, const Line& last) {
    return static_cast<i128>(first.intercept - middle.intercept) *
               (last.slope - middle.slope) >=
           static_cast<i128>(middle.intercept - last.intercept) *
               (middle.slope - first.slope);
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int segment_count, maximum_x;
    std::cin >> segment_count >> maximum_x;

    // Đọc từng đoạn (0, left_y) - (m, right_y) và quy về đường thẳng:
    // slope = (right_y - left_y) / m (chia hết vì slope là số nguyên),
    // intercept = left_y (giá trị tại x = 0).
    std::vector<Line> lines;
    lines.reserve(static_cast<std::size_t>(segment_count));
    for (int index = 0; index < segment_count; ++index) {
        long long left_y, right_y;
        std::cin >> left_y >> right_y;
        lines.push_back({
            (right_y - left_y) / maximum_x,
            left_y
        });
    }

    // Sắp xếp theo slope tăng dần; cùng slope thì để intercept lớn nhất lên trước.
    std::sort(lines.begin(), lines.end(), [](const Line& first,
                                               const Line& second) {
        if (first.slope != second.slope) {
            return first.slope < second.slope;
        }
        return first.intercept > second.intercept;
    });

    // Xây upper hull bằng stack (Convex Hull Trick đơn điệu).
    std::vector<Line> hull;
    hull.reserve(lines.size());
    for (const Line& line : lines) {
        // Cùng slope: chỉ giữ đường có intercept lớn nhất (đã ở đầu nhóm).
        if (!hull.empty() && hull.back().slope == line.slope) {
            continue;
        }
        // Loại các đường ở đỉnh không còn nằm trên bao trên.
        while (hull.size() >= 2 &&
               redundant(hull[hull.size() - 2], hull.back(), line)) {
            hull.pop_back();
        }
        hull.push_back(line);
    }

    // Truy vấn x = 0..m tăng dần: chỉ số đường tối ưu không giảm nên dùng con trỏ tiến.
    std::size_t best = 0;
    for (int x = 0; x <= maximum_x; ++x) {
        while (best + 1 < hull.size() &&
               hull[best + 1].value(x) >= hull[best].value(x)) {
            ++best;
        }
        if (x > 0) {
            std::cout << ' ';
        }
        std::cout << hull[best].value(x);
    }
    std::cout << '\n';
    return 0;
}
